pkg/operator/operators/clustermanager/controllers/certrotationcontroller/certrotation_controller.go

```diff
@@ -239,21 +239,34 @@ func (c certRotationController) syncOne(ctx context.Context, clustermanager *ope
 		}
 	}
 
+	var errs []error
 	// Ensure certificates are exists
 	rotations := c.rotationMap[clustermanagerName]
 
-	if helpers.GRPCAuthEnabled(clustermanager) && !hasRotation(rotations.targetRotations, helpers.GRPCServerSecret) {
+	if helpers.GRPCAuthEnabled(clustermanager) {
 		// maintain the grpc serving certs
 		// TODO may support user provided certs
-		rotations.targetRotations = append(rotations.targetRotations, certrotation.TargetRotation{
-			Namespace: clustermanagerNamespace,
-			Name:      helpers.GRPCServerSecret,
-			Validity:  TargetCertValidity,
-			HostNames: helpers.GRPCServerHostNames(clustermanagerNamespace, clustermanager),
-			Lister:    c.secretInformers[helpers.GRPCServerSecret].Lister(),
-			Client:    c.kubeClient.CoreV1(),
-		})
-		c.rotationMap[clustermanagerName] = rotations
+		hostNames, grpcErr := helpers.GRPCServerHostNames(c.kubeClient, clustermanagerNamespace, clustermanager)
+		if grpcErr != nil {
+			errs = append(errs, grpcErr)
+		} else if hasRotation(rotations.targetRotations, helpers.GRPCServerSecret) {
```

Comment:
could this hasRotation return the rotation so we do not need another for loop?

Reply:
looking at the code, I think changing targetRotations to a map will make it much simpler.